content/en/blog/_posts/2024-04-01-validating-admission-policy-ga/index.md

```diff
@@ -0,0 +1,216 @@
+```shell
+kubectl get -oyaml validatingadmissionpolicies/pod-security.policy.example.com
+```
+```yaml
+  status:
+    typeChecking:
+      expressionWarnings:
+      - fieldRef: spec.validations[3].expression
+        warning: |
+          apps/v1, Kind=Deployment: ERROR: <input>:1:76: undefined field 'Privileged'
+           | object.spec.template.spec.containers.all(c, !has(c.securityContext) || !has(c.securityContext.Privileged) || !c.securityContext.Privileged)
+           | ...........................................................................^
+          ERROR: <input>:1:128: undefined field 'Privileged'
+           | object.spec.template.spec.containers.all(c, !has(c.securityContext) || !has(c.securityContext.Privileged) || !c.securityContext.Privileged)
+           | ...............................................................................................................................^
+
+```
+The policy was checked against its matched type, which is `apps/v1.Deployment`.
+Looking at the `fieldRef`, the problem was with the 3rd expression (index starts with 0)
+The expression in question accessed an undefined `Privileged` field.
+Ahh, looks like it was a copy-and-paste error. The field name should be in lowercase.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "pod-security.policy.example.com"
+spec:
+  failurePolicy: Fail
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  validations:
+  - expression: object.spec.template.spec.containers.all(c, has(c.securityContext) && has(c.securityContext.runAsNonRoot) && c.securityContext.runAsNonRoot)
+    message: 'all containers must set runAsNonRoot to true'
+  - expression: object.spec.template.spec.containers.all(c, has(c.securityContext) && has(c.securityContext.readOnlyRootFilesystem) && c.securityContext.readOnlyRootFilesystem)
+    message: 'all containers must set readOnlyRootFilesystem to true'
+  - expression: object.spec.template.spec.containers.all(c, !has(c.securityContext) || !has(c.securityContext.allowPrivilegeEscalation) || !c.securityContext.allowPrivilegeEscalation)
+    message: 'all containers must NOT set allowPrivilegeEscalation to true'
+  - expression: object.spec.template.spec.containers.all(c, !has(c.securityContext) || !has(c.securityContext.privileged) || !c.securityContext.privileged)
+    message: 'all containers must NOT set privileged to true'
+```
+Check its status again, and you should see all warnings cleared.
+
+Next, let's create a namespace for our tests.
+```shell
+kubectl create namespace policy-test
+```
+Then, bind the policy to the namespace. But at this point, we set the action to `Warn` 
+so that the policy prints out warnings instead of rejecting the requests.
```

Comment:
Optional, but recommended:
```suggestion
so that the policy prints out [warnings](/blog/2020/09/03/warnings/) instead of rejecting the requests.
```